### PR TITLE
Removed CondaRuntimeError and RuntimeError

### LINKED ISSUE #5012: remove instances of RuntimeError and CondaRuntimeError
epitome of tech debt

### PR DESCRIPTION
**Pull request by tyler-thetyrant**
closes #5012

### CODE REVIEW

File: conda/_vendor/auxlib/factory.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 """
 from __future__ import absolute_import, division, print_function
 
+from conda.exceptions import CondaVerificationError
```
Comment (kalefranz) [inline conda/_vendor/auxlib/factory.py]: Use relative imports.  Actually just add this to line 8 below.

Comment (kalefranz) [inline conda/_vendor/auxlib/factory.py]: Actually, undo the changes in `_vendor`.  Those shouldn't be importing from `conda` at all.

File: conda/_vendor/progressbar/__init__.py [hunk 2]
```diff
@@ -274,7 +276,7 @@ def update(self, value=None):
 
         if not self._need_update(): return
         if self.start_time is None:
-            raise RuntimeError('You must call "start" before calling "update"')
+            raise ArgumentError('You must call "start" before calling "update"')
```
Comment (kalefranz) [inline conda/_vendor/progressbar/__init__.py]: Leave as RuntimeError here too.

File: conda/cli/common.py [hunk 3]
```diff
@@ -405,8 +405,9 @@ def ensure_use_local(args):
     try:
         from conda_build.config import croot  # noqa
     except ImportError as e:
-        raise CondaRuntimeError("%s: you need to have 'conda-build >= 1.7.1' installed"
-                                " to use the --use-local option." % e)
+        raise CondaDependencyError("%s: you need to have 'conda-build >= 1.7.1' installed"
+                                   " to use the --use-local option." % e)
+
```
Comment (kalefranz) [inline conda/cli/common.py]: Remove this whole try/except block altogether.  Cut lines 405-410.

File: conda/cli/install.py [hunk 4]
```diff
@@ -21,11 +21,10 @@
...
 from ..core.linked_data import linked as install_linked
-from ..exceptions import (CondaEnvironmentNotFoundError,
-                          CondaIOError, CondaImportError, CondaOSError,
-                          CondaRuntimeError, CondaSystemExit, CondaValueError,
-                          DirectoryNotFoundError, DryRunExit, LockError, NoPackagesFoundError,
-                          PackageNotFoundError, TooManyArgumentsError, UnsatisfiableError)
+from ..exceptions import CondaEnvironmentNotFoundError, CondaHistoryError, CondaIOError, \
+    CondaImportError, CondaOSError, CondaSystemExit, CondaValueError, DirectoryNotFoundError, \
+    DryRunExit, LockError, NoPackagesFoundError, PackageNotFoundError, TooManyArgumentsError, \
+    UnsatisfiableError
```
Comment (kalefranz) [inline conda/cli/install.py]: cmd-option-o does the `\` thing.  Switch it back to using parenthesis.

File: conda/cli/install.py [hunk 5]
```diff
@@ -367,11 +366,11 @@ def install(args, parser, command='install'):
                         else:
                             raise CondaIOError("Can't write the history file", e)
 
-            except RuntimeError as e:
+            except LockError as e:
```
Comment (kalefranz) [inline conda/cli/install.py]: I don't think we have LockError used anywhere anymore.  You can probably get rid of it.

File: conda/cli/main_package.py [hunk 6]
```diff
@@ -21,6 +21,7 @@
 from conda.core.linked_data import linked
 from os.path import basename, dirname, isfile, islink, join, abspath, isdir
 
+from conda.exceptions import CondaVerificationError
```
Comment (kalefranz) [inline conda/cli/main_package.py]: Use relative import.

File: conda/core/linked_data.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
 from os import listdir
 from os.path import isdir, isfile, join
 
+from conda.exceptions import CondaDependencyError
```
Comment (kalefranz) [inline conda/core/linked_data.py]: Use relative import.

File: conda/core/portability.py [hunk 8]
```diff
@@ -60,7 +60,7 @@ def _update_prefix(original_data):
                           'new_prefix': new_prefix,
                           'placeholder': placeholder,
                           })
-            raise CondaRuntimeError(message)
+            raise CondaRevisionError(message)
```
Comment (kalefranz) [inline conda/core/portability.py]: Just create a new exception here.  Call it `PrefixReplacementError`.

Comment (kalefranz) [inline conda/core/portability.py]:
```
raise PrefixReplacementError(new_data_length=len(data), original_data_lengh=len(original_data), ...)
```

File: conda/core/repodata.py [hunk 9]
```diff
@@ -64,7 +64,7 @@ def collect_all_repodata(use_cache, tasks):
             import concurrent.futures
             executor = concurrent.futures.ThreadPoolExecutor(10)
             repodatas = _collect_repodatas_concurrent(executor, use_cache, tasks)
-        except (ImportError, RuntimeError) as e:
+        except ImportError as e:
             # concurrent.futures is only available in Python >= 3.2 or if futures is installed
             # RuntimeError is thrown if number of threads are limited by OS
```
Comment (kalefranz) [inline conda/core/repodata.py]: Catching a `RuntimeError` from another library is fine.  We just don't want to be raising `RuntimeError` in conda code.  Per the comment here, we actually need to leave this.

File: conda/exceptions.py [hunk 10]
```diff
@@ -575,7 +576,7 @@ def conda_exception_handler(func, *args, **kwargs):
             return return_value
     except CondaExitZero:
         return 0
-    except CondaRuntimeError as e:
+    except CondaIOError as e:
         print_unexpected_error_message(e)
         return 1
     except CondaError as e:
```
Comment (kalefranz) [inline conda/exceptions.py]: We should be able to remove lines 579-581 altogether now.

File: conda/misc.py [hunk 11]
```diff
@@ -259,7 +259,7 @@ def clone_env(prefix1, prefix2, verbose=True, quiet=False, index_args=None):
         what = "Package%s " % ('' if len(notfound) == 1 else 's')
         notfound = '\n'.join(' - ' + fn for fn in notfound)
         msg = '%s missing in current %s channels:%s' % (what, context.subdir, notfound)
-        raise CondaRuntimeError(msg)
+        raise ChannelError(msg)
```
Comment (kalefranz) [inline conda/misc.py]: Use `PackageNotFoundError` here.  And fix up the other uses of that exception so they all have a coherent error message.  Use `location` as an input argument to the exception, where location is either usually a prefix or list of channels it looks like.

File: conda/version.py [hunk 12]
```diff
@@ -341,7 +341,7 @@ def __new__(cls, spec):
         elif spec.startswith(('=', '<', '>', '!')):
             m = version_relation_re.match(spec)
             if m is None:
-                raise CondaRuntimeError('Invalid version spec: %s' % spec)
+                raise CondaDependencyError('Invalid version spec: %s' % spec)
```
Comment (kalefranz) [inline conda/version.py]: We're updating the MatchSpec object in another pull request.  Just make this `CondaError` for now.

Comment (kalefranz) [inline conda/version.py]: Actually, go ahead and create a new exception here.

```
class InvalidSpecError(CondaError):

    def __init__(self, invalid_spec):
        message = "Invalid spec: %(invalid_spec)s"
        super(InvalidSpecError, self).__init__(message, invalid_spec=invalid_spec)
```

Comment (kalefranz) [inline conda/version.py]: Also use this for any place in `Version`, `VersionSpec`, or `MatchSpec` that looks like makes sense.

File: conda/core/portability.py [hunk 13]
```diff
@@ -49,18 +49,9 @@ def _update_prefix(original_data):
...
-                       % {'new_data_length': len(data),
-                          'original_data_length': len(original_data),
-                          'path': path,
-                          'new_prefix': new_prefix,
-                          'placeholder': placeholder,
-                          })
-            raise CondaRuntimeError(message)
+            message = "PrefixReplacementError"
+            raise PrefixReplacementError(message, len(original_data), len(data), new_prefix,
+                                         path, placeholder)
```
Comment (kalefranz) [inline conda/core/portability.py]: `message` doesn't do anything useful here I don't think.  Just take it out.

File: conda/exceptions.py [hunk 14]
```diff
@@ -486,6 +481,36 @@ def __init__(self, message):
...
+        super(CondaDependencyError, self).__init__(message)
+
+
+class PrefixReplacementError(CondaError):
+    def __init__(self, message, new_data_length, original_data_length, new_prefix,
+                 path, placeholder):
+        message = dals("""
+        Refusing to replace data of length '%(new_data_length)d' with
+        data of length '%(original_data_length)d' for binary file
+        New Data Length:      %(new_data_length)d
```
Comment (kalefranz) [inline conda/exceptions.py]: Lower case "new data length" and indent two spaces.  Only use one space after colon.  Make formatting of parameters below similar.

File: conda/exceptions.py [hunk 15]
```diff
@@ -486,6 +481,36 @@ def __init__(self, message):
         super(CondaVerificationError, self).__init__(message)
 
 
+class CondaDependencyError(CondaError):
+    def __init__(self, message):
+        super(CondaDependencyError, self).__init__(message)
+
+
+class PrefixReplacementError(CondaError):
+    def __init__(self, message, new_data_length, original_data_length, new_prefix,
```
Comment (kalefranz) [inline conda/exceptions.py]: Don't need the first `message` argument here.  Need it in the `super()` call of course.

Comment (kalefranz) [inline conda/exceptions.py]: Also, let's rename this to `BinaryPrefixReplacementError` because it's specific to binary files.

File: conda/core/repodata.py [hunk 16]
```diff
@@ -64,7 +64,7 @@ def collect_all_repodata(use_cache, tasks):
             import concurrent.futures
             executor = concurrent.futures.ThreadPoolExecutor(10)
             repodatas = _collect_repodatas_concurrent(executor, use_cache, tasks)
-        except (ImportError, RuntimeError) as e:
+        except (ImportError, RuntimeError)as e:
```
Comment (kalefranz) [inline conda/core/repodata.py]: Missing a space after `)`

File: conda/misc.py [hunk 17]
```diff
@@ -259,7 +259,7 @@ def clone_env(prefix1, prefix2, verbose=True, quiet=False, index_args=None):
         what = "Package%s " % ('' if len(notfound) == 1 else 's')
         notfound = '\n'.join(' - ' + fn for fn in notfound)
         msg = '%s missing in current %s channels:%s' % (what, context.subdir, notfound)
-        raise CondaRuntimeError(msg)
+        raise PackageNotFoundError(msg)
```
Comment (kalefranz) [inline conda/misc.py]: We still need to clean up this error now.  Most all of this message construction should be done inside the exception class itself.  Looking at other uses of this error, you'll need to use `location `as an input argument to the exception, where location is either usually a prefix or list of channels it looks like.

**Comment by kalefranz** [comment]
CC @goanpeca

**Comment by goanpeca** [comment]
Hi @tyler-thetyrant let's create an issue associated to the work in this PR, so we start organizing things better :-)
